### PR TITLE
Track ref returns for Make Field Readonly

### LINKED ISSUE #33009: IDE0044 "Make field readonly" false positive when used in ref return
**Version Used**: 
16.0.0 Preview 2.0

**Steps to Reproduce**:
```csharp
public class C
{
    private int a, b; // IDE0044
    private ref int select(bool c) => ref (c ? ref a : ref b);
    public void Test() => select(true) = 3;
}
```

**Expected Behavior**:
The members are mutated. Marking them readonly breaks the code, so IDE0044 shouldn't trigger.

**Actual Behavior**:
IDE0044 is falsely triggered.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #33009

### CODE REVIEW

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs [hunk 1]
```diff
@@ -130,6 +130,32 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
...
+                    default:
+                        return ValueUsageInfo.Read;
+                }
+            }
+            else if (operation.Parent is IConditionalOperation conditionalOperation)
+            {
+                if (operation == conditionalOperation.WhenTrue
+                    || operation == conditionalOperation.WhenFalse)
+                {
+                    return GetValueUsageInfo(conditionalOperation);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: is this true? would it depend on how the conditional was used *and* if 'ref' was used inside the conditional?  (i'm honeslty not sure).

Comment (sharwell) [reply src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: The return type of the method determines the behavior here.

* If the method returns by-value or by-ref-readonly, the branches of the conditional will be treated as readonly references
* If the method returns by-ref, the branches of the conditional will be `ref` and failure to include the `ref` around the whole conditional would be a compilation error

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs [hunk 2]
```diff
@@ -152,6 +152,29 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
                         return ValueUsageInfo.Read;
                 }
             }
+            else if (operation.Parent is IReturnOperation)
+            {
+                var containingMethod = TryGetContainingAnonymousFunctionOrLocalFunction(operation)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs:157]: @mavasani can we not find the containing method from the op-tree?  should we be able to get that?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs [hunk 3]
```diff
@@ -185,6 +208,26 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
             return ValueUsageInfo.Read;
         }
 
+        public static IMethodSymbol TryGetContainingAnonymousFunctionOrLocalFunction(this IOperation operation)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs:211]: NRT-ise this  file?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs (hunk 2 above)
Comment (mavasani) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs:157]: Nope, that information is not present in the Operation tree.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/OperationExtensions.cs:211]: I updated it locally, but will submit that change separately to master branch.

**Pull request merged**